### DISCUSSION TITLE
How to typehint an attribute of a generic type?

### DISCUSSION THREAD

**Discussion by FeryET** [python/typing#1510] [Q&A] [answered]
Hi.

I'm trying to write a decorator for my functions that gets the function result type as its input, and then returns only the required attribute.

Something similar to this idea, look at the `Attribute` typing class that I have used. 

```python

Params = ParamSpec("Params")
T = TypeVar("T")

def dedecorator(key: str):
      def decorator(func: Callable[Params, T]) -> Callable[Params, Attribute[T, key]]:
             def wrapper(*args: Params.args, **kwargs: Params.kwargs) -> Attribute[T, key]:
                    return getattr(func(*args, **kwargs), key)
             return wrapper
      return decorator
```

Is something possible in Python as of now?

**Comment by erictraut** [comment, marked as answer]
No, that's not possible in the type system today. If there's a specific attribute that you're interested in, you could define a generic protocol class, like this:

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=FAMwTg9gtgBALgTwA4EsB2BzGKpImOGAYQEMAbMkgIzIFMAaGABUjggGMIzgf3KBnfjADKAVyR4C-AGIQIAbQAqAXQAULCG05kAlAC5gMIzABEZkXDCj2cUWHLxktGCHwwSMPiUHwAFiUJ%2BcUk4IQ8wWhIAEwB6AHcwFDhnE1cIExgkEnsoWmSwUzNDYzS9GEUeKNoQGCrOezYweQAqZqZskih%2BRhVVEFE0djLSCmo6eXac7pFg-FDZBRVlHRgAWgA%2BYnJKGloJjq6e5QNjWuqYBJIJWjBVZuyMfjLJzv4AOgfp1oBrOM-ng7vX6fFYbconU7GCK2MBoGAYPIBSx9AbsO6fRg-P5gR46RgAcjS%2BJ0xSM0LscMu1zAPGAXh8CwAkrgyBCjKVsGg4DwAAJ1fABfDAKo1fqDVSgzZMllsmBveXAIA)

```python
from typing import Callable, Protocol

class SupportsFoo[T](Protocol):
    """ Structural type for a class that supports a read/write "foo" parameter """
    foo: T

def decorator[**Params, T](func: Callable[Params, SupportsFoo[T]]) -> Callable[Params, T]:
    def wrapper(*args: Params.args, **kwargs: Params.kwargs) -> T:
        return getattr(func(*args, **kwargs), 'foo')
    return wrapper


class FooImpl:
    foo: int

@decorator
def func() -> FooImpl:
    ...
```

**Comment by Daverball** [comment]
There is also a mypy plugin that may help you get there https://pypi.org/project/mypy-extras/ which has an `AttrOf` special type, although it expects a `Literal` for the second argument, so it probably wouldn't work, unless you also wrote a mypy plugin specifically for this decorator, in order to resolve the first argument to a specific `Literal` instead of `str` (maybe using `LiteralString` instead of `str` would accomplish this, although I'm not sure it actually would).
